File: EmotionEngine/src/main/java/co/q64/emotion/lang/value/Value.java
package co.q64.emotion.lang.value;

import co.q64.emotion.types.Operation;
import co.q64.emotion.util.Rational;

import java.util.List;

public interface Value extends Comparable<Value> {
    public Value operate(Value value, Operation type);

    public int compareTo(Value value);

    public List<Value> iterate();

    public Rational asNumber();

    public boolean asBoolean();

    public default boolean isBoolean() {
        return false;
    }

    public default boolean isList() {
        return false;
    }

    public default boolean isNumber() {
        return false;
    }

    public default int asInt() {
        return asNumber().intValue();
    }

    public default long asLong() {
        return asNumber().longValue();
    }

    public default double asDouble() {
        return asNumber().doubleValue();
    }

    public default char asChar() {
        if (toString().length() > 0) {
            return toString().toCharArray()[0];
        }
        return 0;
    }
}

File: EmotionEngine/src/main/kotlin/co/q64/emotion/lang/value/Value.kt
package co.q64.emotion.lang.value

import co.q64.emotion.lang.Operation
import co.q64.emotion.math.Rational
import co.q64.emotion.math.rational

interface Value : Comparable<Value> {
    fun operate(other: Value, type: Operation): Value
    val list: List<Value>
    val number: Rational
    val int: Int get() = number.intValue()
    val long: Long get() = number.longValue()
    val double: Double get() = number.doubleValue()
    val boolean: Boolean get() = number != Rational.ZERO
    val text: String get() = toString()
}

fun value(value: String): Value {
    TODO()
}
fun Rational.value(): Value = NumberValue(this)
fun Int.value(value: Int): Value = rational().value()
fun Long.value(value: Long): Value = rational().value()
fun Double.value(): Value = rational().value()
fun Value.value() = if (this is TextValue) value(this.text) else this